Core/include/Acts/Utilities/Intersection.hpp

```diff
@@ -39,144 +39,210 @@ inline std::ostream& operator<<(std::ostream& os, IntersectionStatus status) {
 template <typename object_t, typename representation_t = object_t>
 class ObjectIntersection {
  public:
+  /// Default constructor
+  ObjectIntersection() = default;
+
+  /// Object intersection - symmetric setup
+  ///
+  /// @param intersection is the intersection
+  /// @param object is the object to be instersected
+  /// @param index is the intersection index
+  template <typename T = representation_t,
+            std::enable_if_t<std::is_same<T, object_t>::value, int> = 0>
+  ObjectIntersection(const Intersection3D& intersection, const object_t* object,
+                     std::uint8_t index = 0)
+      : m_intersection(intersection),
+        m_object(object),
+        m_representation(object),
+        m_index(index) {}
+
+  /// Object intersection
+  ///
+  /// @param intersection is the intersection
+  /// @param object is the object to be instersected
+  /// @param representation is the object representation
+  /// @param index is the intersection index
+  ObjectIntersection(const Intersection3D& intersection, const object_t* object,
+                     const representation_t* representation,
+                     std::uint8_t index = 0)
+      : m_intersection(intersection),
+        m_object(object),
+        m_representation(representation),
+        m_index(index) {}
+
+  /// Returns whether the intersection was successful or not
+  explicit operator bool() const { return m_intersection.operator bool(); }
+
+  const Intersection3D& intersection() const { return m_intersection; }
+
+  const Intersection3D::Position& position() const {
+    return m_intersection.position();
+  }
+
+  ActsScalar pathLength() const { return m_intersection.pathLength(); }
+
+  Intersection3D::Status status() const { return m_intersection.status(); }
+
+  const object_t* object() const { return m_object; }
+
+  const representation_t* representation() const { return m_representation; }
+
+  std::uint8_t index() const { return m_index; }
+
+  static bool forwardOrder(const ObjectIntersection& aIntersection,
+                           const ObjectIntersection& bIntersection) {
+    return Intersection3D::forwardOrder(aIntersection.intersection(),
+                                        bIntersection.intersection());
+  }
+
+  static bool closestOrder(const ObjectIntersection& aIntersection,
+                           const ObjectIntersection& bIntersection) {
+    return Intersection3D::closestOrder(aIntersection.intersection(),
+                                        bIntersection.intersection());
+  }
+
+ private:
   /// The intersection itself
-  Intersection3D intersection{};
+  Intersection3D m_intersection;
   /// The object that was (tried to be) intersected
-  const object_t* object{nullptr};
+  const object_t* m_object = nullptr;
   /// The representation of this object
-  const representation_t* representation{nullptr};
+  const representation_t* m_representation = nullptr;
+  /// The intersection index
+  std::uint8_t m_index = 0;
```

Comment:
Should this be part of the intersection object itself?

Reply:
I think my motivation was that the context of the object will tell you why an index is necessary and the intersection can stay context-less this way

Reply:
ok